pkg/util/utils.go

```diff
@@ -374,53 +385,49 @@ func UpdateK8sResourceAnnotation(client client.Client, ctx *context.Context, k8s
 	return nil
 }
 
-// GenerateID generate id for nsx resource, some resources has complex index, so set it type to string
-func GenerateID(res_id, prefix, suffix string, index string) string {
-	var id strings.Builder
-	if len(prefix) > 0 {
-		id.WriteString(prefix)
-		id.WriteString("_")
-	}
-
-	id.WriteString(res_id)
-	if len(index) > 0 {
-		id.WriteString("_")
-		id.WriteString(index)
+// GenerateIDByObject generate string id for nsx resource using the provided Object's name and uid. Note,
+// this function is used on the resources with VPC scenario, and the provided obj is the K8s CR which is
+// used to generate the NSX resource.
+func GenerateIDByObject(obj metav1.Object) string {
+	return normalizeNameByLimit(obj.GetName(), string(obj.GetUID()), common.MaxIdLength)
+}
 
+// GenerateIDByObjectByLimit generate string id for nsx resource using the provided Object's name and uid,
+// and truncate the string with the given limit length.
+func GenerateIDByObjectByLimit(obj metav1.Object, limit int) string {
```

Comment:
Do you think we can make GenerateIDByObject and GenerateIDByObjectByLimit unified, only use GenerateIDByObjectByLimit, seemly less code maintenance.

Reply:
I would prefer to leave function `GenerateIDByObject` which has a single parameter of the CR object (or the ObjectMeta). In this function, we would hide the length limit (internal limit from nsx-operator but not from NSX) inside the function implementation, then we can keep the caller to be insensitive with the id length. In this way, we can easily modify the id length limit in the future if any changes are proposed from NSX side in the future, and we don't need to modify every caller.  

As for function `GenerateIDByObjectByLimit`, it is mainly used in the case that an explicitly length limit exists, like Subnet's display_name, or VPC/LBS display_name.